README.md
---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->



# PathInterpolatR

The package PathInterpolatR provides access to some simple functions for path interpolation (e.g., from GPS tracking data). Most notably, it is associated with the methods described in:

Long, J.A. (2016) Kinematic interpolation of movement data. International Journal of
Geographical Information Science. 30(5): 854-868.

## Installation

You can install PathInterpolatR from github with:


```r
# install.packages("devtools")
devtools::install_github("jedalong/PathInterpolatR")
```

## Quick Demo

Below a quick demo shows how the package can be used with the kinematic interpolation function.


```r
library(PathInterpolatR)
data(contrived)
xyt <- contrived
###times for interpolation
t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
### add velocities for kinematic analysis
xytvv <- cbind(xyt[2:3,],rbind(c(0,3),c(3,0)))
a <- kinematic(xytvv,t.slice)
plot(xyt[,1],xyt[,2],pch=20,asp=1)
points(a[,1],a[,2])
```

![plot of chunk example](README-example-1.png)

--- End ---


NEWS
PathInterpolatr 0.1
=================
* 23-09-2015 -  First Beta version developed.


R/bezier.r
# ---- roxygen documentation ----
#
#' @title Interpolate using Bezier curve
#'
#' @description
#'  Perform path interpolation using a Bezier curve interpolation method.
#'
#' @details
#'  Bezier curves can be useful for interpolating movement when the object exhibits curvi-linear movement shapes (Long 2015). For example, the movements of marine mammals often exhibit this property (Tremblay \emph{et al.} 2006) . The Bezier method, as implemented here, requires four control points (2-D coordinates) as input, along with their corresponding times; these four points are passed to the function as a dataframe with four rows, and three columns corresponding to x, y, & t (the names of the columns do not matter). The function interpolates the movement trajectory for the times indicated by parameter \code{t.slice}. The \code{t.slice} times must lie between the time of the 2nd and 3rd control point. Times can be passed in as either POSIX-class objects or numeric, but not a mixture of both.
#'
#' @param xyt a 4x3 dataframe containing the coordinates and times for the four control points. Each row of the dataframe should be arranged as x, y, t.
#' @param t.slice a single time (POSIX or numeric), or list of times, to be interpolated for. The times must lie between those of the middle two control points (i.e., the times associated with rows 2 & 3 in the \code{xyt} dataframe).
#'
#' @return
#'  The function returns a dataframe (with \code{nrow = length{t.slice}}) corresponding to the interpolated locations.
#'
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. \cr \cr
#' Tremblay, YC \emph{et al.} (2006) Interpolation of animal tracking data in a fluid environment. \emph{Journal of Experimental Biology}. 209(1): 128-140. 
#'
#' @keywords interpolation
#'  
#' @examples
#' data(contrived)
#' xyt <- contrived
#' ###times for interpolation
#' t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
#' a <- bezier(xyt,t.slice)
#' plot(xyt[,1],xyt[,2],pch=20)
#' points(a[,1],a[,2])
#'
#' @export
#
# ---- End of roxygen documentation ----

bezier <- function(xyt,t.slice){
  if (dim(xyt)[1] != 4){stop('The xyt dataframe does not contain 4 rows and 3 columns.')}
  if (dim(xyt)[2] != 3){stop('The xyt dataframe does not contain 4 rows and 3 columns.')}
  if (min(t.slice) < xyt[2,3]){stop('Invalid interpolation times in t.slice.')}
  if (max(t.slice) > xyt[3,3]){stop('Invalid interpolation times in t.slice.')}
  
  nameo <- names(xyt)
  xyt <- as.matrix(xyt)
  
  x1 <- xyt[1,1:2]
  x2 <- xyt[2,1:2]
  x3 <- xyt[3,1:2]
  x4 <- xyt[4,1:2]
  
  t1 <- xyt[1,3]
  t2 <- xyt[2,3]
  t3 <- xyt[3,3]
  t4 <- xyt[4,3]
  dt <- as.numeric(t3-t2,units='secs')
  
  v1 <- (x2-x1)/(t2-t1)
  v2 <- (x4-x3)/(t4-t3)
  v2 <- -v2
  
  p0 <- x2
  p1 <- x2+v1*0.5*dt
  p2 <- x3+v2*0.5*dt
  p3 <- x3
  
  t <- as.numeric(t.slice - t2,units='secs') / as.numeric(dt,units='secs')
  bx <- (1-t)^3*p0[1] + 3*(1-t)^2*t*p1[1]+3*(1-t)*t^2*p2[1]+t^3*p3[1]
  by <- (1-t)^3*p0[2] + 3*(1-t)^2*t*p1[2]+3*(1-t)*t^2*p2[2]+t^3*p3[2]
  xy <- data.frame(x=bx,y=by,t=t.slice,row.names=NULL)
  names(xy) <- nameo
  return(xy)
}





R/linear.R
# ---- roxygen documentation ----
#
#' @title Interpolate using linear interpolation
#'
#' @description
#'  Perform linear path interpolation on a movement dataset.
#'
#' @details
#'  Linear interpolation is the simplest, and most commonly employed methods for path interpolation. 
#'
#' @param xyt a 2x3 dataframe containing the coordinates and times of the two points to be interpolated between, often termed the anchor points. Each row of the dataframe should be arranged as x, y, t.
#' @param t.slice a single time (POSIX or numeric), or list of times, to be interpolated for. The times must lie between those of the points in \code{xyt}.
#'
#' @return
#'  The function returns a dataframe (with \code{nrow = length{t.slice}}) corresponding to the interpolated locations.
#'
#' @keywords interpolation
#' @examples
#' data(contrived)
#' xyt <- contrived
#' ###times for interpolation
#' t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
#' a <- linear(xyt[2:3,],t.slice)
#' plot(xyt[,1],xyt[,2],pch=20)
#' points(a[,1],a[,2])
#' 
#' @export
#
# ---- End of roxygen documentation ----
linear <- function(xyt,t.slice){
  if (dim(xyt)[1] != 2){stop('The xyt dataframe does not contain 2 rows and 3 columns.')}
  if (dim(xyt)[2] != 3){stop('The xyt dataframe does not contain 2 rows and 3 columns.')}
  if (min(t.slice) < xyt[1,3]){stop('Invalid interpolation time(s) in t.slice.')}
  if (max(t.slice) > xyt[2,3]){stop('Invalid interpolation time(s) in t.slice.')}
  
  nameo <- names(xyt)
  xyt <- as.matrix(xyt)
  
  x1 <- xyt[1,1:2]
  x2 <- xyt[2,1:2]
  t1 <- xyt[1,3]
  t2 <- xyt[2,3]
  
  t <- t2 - t1
  t.s <- t.slice-t1
  rat <- t.s/t
  bx <-  rat*(x2[1]-x1[1]) + x1[1]
  by <- rat*(x2[2]-x1[2]) + x1[2]
  xy <- data.frame(x=bx,y=by,t=t.slice)
  names(xy) <- nameo
  return(xy)
}

R/splineCR.R
# ---- roxygen documentation ----
#
#' @title Interpolate using Catmull-Rom spline
#'
#' @description
#'  Perform path interpolation using the Catumull-Rom spline interpolation method.
#'
#' @details
#'  Catmull-Rom splines can be useful for interpolating movement when the object exhibits curvi-linear movement shapes (Long 2015). For example, hurricanes often exhibit this property. The Catmull-Rom method requires four control points (2-D coordinates) as input, along with their corresponding times (Barry & Goldman 1988); these four points are passed to the function as a dataframe with four rows, and three columns corresponding to x, y, & t (the names of the columns do not matter). The function interpolates the movement trajectory for the times indicated by parameter \code{t.slice}. The \code{t.slice} times must lie between the time of the 2nd and 3rd control point. Times can be passed in as either POSIX-class objects or numeric, but not a mixture of both.
#'
#' @param xyt a 4x3 dataframe containing the coordinates and times for the four control points. Each row of the dataframe should be arranged as x, y, t.
#' @param t.slice a single time (POSIX or numeric), or list of times, to be interpolated for. The times must lie between those of the middle two control points (i.e., the times associated with rows 2 & 3 in the \code{xyt} dataframe).
#'
#' @return
#'  The function returns a dataframe (with \code{nrow = length{t.slice}}) corresponding to the interpolated locations.
#'
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. \cr \cr
#' Barry, PJ, Goldman, RN (1988) A recursive evaluation algorithm for a class of Catmull-Rom splines. \emph{ACM SIGGRAPH Computer Graphics}. 22(4): 199-204.
#' 
#' @keywords interpolation
#' @examples
#' data(contrived)
#' xyt <- contrived
#' ###times for interpolation
#' t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
#' a <- splineCR(xyt,t.slice)
#' plot(xyt[,1],xyt[,2],pch=20)
#' points(a[,1],a[,2])
#' 
#' @export
#
# ---- End of roxygen documentation ----

#Catmull-Rom spline
#==============================================================================
splineCR <- function(xyt,t.slice){
  if (dim(xyt)[1] != 4){stop('The xyt dataframe does not contain 4 rows and 3 columns.')}
  if (dim(xyt)[2] != 3){stop('The xyt dataframe does not contain 4 rows and 3 columns.')}
  if (min(t.slice) < xyt[2,3]){stop('Invalid interpolation times in t.slice.')}
  if (max(t.slice) > xyt[3,3]){stop('Invalid interpolation times in t.slice.')}
  
  nameo <- names(xyt)
  xyt <- as.matrix(xyt)
  
  P0 <- xyt[1,1:2]
  P1 <- xyt[2,1:2]
  P2 <- xyt[3,1:2]
  P3 <- xyt[4,1:2]
  
  s0 <- xyt[1,3]
  s1 <- xyt[2,3]
  s2 <- xyt[3,3]
  s3 <- xyt[4,3]
  
  n <- length(t.slice)
  df <- data.frame(x=0,y=0,t=t.slice)  
  for (i in 1:n){
    s. <- t.slice[i]
    A1 <- (s1-s.)/(s1-s0)*P0 + (s.-s0)/(s1-s0)*P1
    A2 <- (s2-s.)/(s2-s1)*P1 + (s.-s1)/(s2-s1)*P2
    A3 <- (s3-s.)/(s3-s2)*P2 + (s.-s2)/(s3-s2)*P3
    
    B1 <- (s2-s.)/(s2-s0)*A1 + (s.-s0)/(s2-s0)*A2
    B2 <- (s3-s.)/(s3-s1)*A2 + (s.-s1)/(s3-s1)*A3
    
    C <- ((s2-s.)/(s2-s1))*B1 + ((s.-s1)/(s2-s1))*B2
    df[i,1:2] <- C
  }  

  names(df) <- nameo
  return(df)

}






R/contrived.R
#' Contrived example dataset
#'
#' A dataset pertaining to the contrived example from Long (2015) used to demonstrate each method for path interpolation.
#' 
#' Contrived example dataset for contrasting different path interpolation methods (open circles) with kinematic interpolation (grey crosses). In the contrived example, the object begins at the point z(0) = (0,-3) with a velocity of 0 m/s and then moves to the origin z(1) = (0, 0) with a velocity of 3 m/s to the North v(1) = (0,3). The object reaches position z(6) = (10,10) after 5 s, it now has a velocity of 3 m/s to the East v(6) = (3,0) and it continues on to location z(7) = (13,10). The time difference between the second and thrid point is 5 s and the object's location is estimated every 0.5 s in between. 
#'
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. 
#' @name contrived  
#' @docType data
#' @keywords datasets
#' @format A dataframe (\code{contrived}) delineating the four anchor points for path interpolation from the contrived example in Long (2015). 
NULL


R/kinematic.R
# ---- roxygen documentation ----
#
#' @title Interpolate using kinematic interpolation
#'
#' @description
#'  Perform kinematic path interpolation on a movement dataset. Kinematic path interpolation was introduced in the paper Long (2015). Kinematic interpolation is appropriate for fast moving objects, recorded with relatively high resolution tracking data.
#'
#' @details
#'  Kinematic interpolation requires the user to input the coordinates of the anchor points between which the interpolation is occurring, as well as initial and final velocities associate with the anchor points. In practice, these velocities may be explicitly known, or estimated from the tracking data. 
#'
#' @param xytvv a 2x5 dataframe containing the coordinates, times, and initial and final velocities (as 2D vectors) of the two points to be interpolated between, often termed the anchor points. Each row of the dataframe should be arranged as x, y, t, vx, vy.
#' @param t.slice a single time (POSIX or numeric), or list of times, to be interpolated for. The times must lie between those of the points in \code{xytv}.
#'
#' @return
#'  The function returns a dataframe (with \code{nrow = length{t.slice}}) corresponding to the interpolated locations.
#'
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. 
#'
#' @keywords interpolation 
#' @examples
#' data(contrived)
#' xyt <- contrived
#' ###times for interpolation
#' t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
#' ### add velocities for kinematic analysis
#' xytvv <- cbind(xyt[2:3,],rbind(c(0,3),c(3,0)))
#' a <- kinematic(xytvv,t.slice)
#' plot(xyt[,1],xyt[,2],pch=20)
#' points(a[,1],a[,2])
#' 
#' @export
#
# ---- End of roxygen documentation ----
kinematic <- function(xytvv,t.slice){
  
  if (dim(xytvv)[1] != 2){stop('The xytvv dataframe does not contain 2 rows and 5 columns.')}
  if (dim(xytvv)[2] != 5){stop('The xytvv dataframe does not contain 2 rows and 5 columns.')}
  if (min(t.slice) < xytvv[1,3]){stop('Invalid interpolation times in t.slice.')}
  if (max(t.slice) > xytvv[2,3]){stop('Invalid interpolation times in t.slice.')}
  
  nameo <- names(xytvv)
  xytvv <- as.matrix(xytvv)
  
  x1 <- xytvv[1,1:2]
  x2 <- xytvv[2,1:2]
  t1 <- xytvv[1,3]
  t2 <- xytvv[2,3]
  v1 <- xytvv[1,4:5]
  v2 <- xytvv[2,4:5]
  
  t <- as.numeric(t2 - t1, units='secs')
  t.s <- as.numeric(t.slice-t1, units='secs')
  
  ax <- matrix(c(t^2/2,t^3/6,t,t^2/2),nrow=2,byrow=T)
  bx <- c(x2[1]-x1[1]-v1[1]*t,v2[1]-v1[1])
  coef.x <- solve(ax,bx)
  
  ay <- ax
  by <- c(x2[2]-x1[2]-v1[2]*t,v2[2]-v1[2])
  coef.y <- solve(ay,by)                                                                                  
  
  pos <- function(t,x1,v1,b,c){x1 + v1*t + (t^2)*b/2 + (t^3)*c/6}
  #vel <- function(t,v1,b,c){v1+b*t+c*t^2/2}
  #acc <- function(t,b,c){b+c*t}
  
  x <- pos(t.s,x1[1],v1[1],coef.x[1],coef.x[2])
  y <- pos(t.s,x1[2],v1[2],coef.y[1],coef.y[2])
  
  xy <- data.frame(x=x,y=y,t=t.slice)
  names(xy) <- nameo[1:3]
  return(xy)
}



R/PathInterpolatR-package.R
# This is package documentation for stampr.
# roxygen will use this file to create a NAMESPACE file.
# Of importance is the @import command, as it lists package dependencies.

#' PathInterpolatR: Methods for Path Interpolation
#'
#' The package \code{PathInterpolatR} provides tools for performing path interpolation on movement data (such as GPS tracking data). These tools were developed in support of the paper Long (2015) and extend the work of other authors (see references within).
#'
#' Currently the code has been developed for the simple scenario of interpolation between two points (where points on either side may or may not be known). Eventually, I will extend the code to facilitate straightforward implementation of interpolation with common movement data structures from other R packages (e.g., \code{adehabitat} and \code{MOVE}).
#'
#' @author Jed Long
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. 
#'
#' @import sp rgeos
#' @docType package
#' @name PathInterpolatR-package
NULL


R/tgrw.r
# ---- roxygen documentation ----
#
#' @title Interpolate using time geographic constrained random walk
#'
#' @description
#'  Perform path interpolation using the constrained random walk method outlined in the paper by Technitis et al. (2015), as implemented in the paper by Long (2015).
#'
#' @details
#'  Many moving objects exhibit movement properties that can be modelled via random walks. Thus, in many cases it is of interest to use random walks as a model for path interpolation. The time geographic constraned random walk is a special case of the random walk, wherby the interpolation is constrained by the space-time prism. The size of the space-time prism is controlled by the parameter \code{vmax}, which sets how far the interpolation is allowed to 'wander'. For more details, please see Technitis et al. (2015).
#'
#' @param xyt a 2x3 dataframe containing the coordinates and times of the two points to be interpolated between, often termed the anchor points of the space-prism. Each row of the dataframe should be arranged as x, y, t.
#' @param t.slice a single time (POSIX or numeric), or list of times, to be interpolated for. The times must lie between those of the points in \code{xyt}.
#' @param vmax parameter controlling the bounds of the constrained random walk. Default value is 1.5 x d/t where d is the distance between the anchor points and t the time difference.
#'
#' @return
#'  The function returns a dataframe (with \code{nrow = length{t.slice}}) corresponding to the interpolated locations.
#'
#' @references
#' Long, JA (2015) Kinematic interpolation of movement data. \emph{International Journal of Geographical Information Science}. DOI: 10.1080/13658816.2015.1081909. \cr \cr
#' Technitis, G. \emph{et al.} (2015) From A to B, randomly: A point-to-point random trajectory generator for animal movement. \emph{International Journal of Geographical Information Science}. 29(6): 912-934. 
#'
#' @keywords interpolation
#' @examples
#' data(contrived)
#' xyt <- contrived
#' ###times for interpolation
#' t.slice <- c(1.5,2,2.5,3,3.5,4,4.5,5,5.5)
#' a <- tgrw(xyt[2:3,],t.slice,vmax=6)
#' plot(xyt[,1],xyt[,2],pch=20)
#' points(a[,1],a[,2])
#' 
#' @export
#
# ---- End of roxygen documentation ----

tgrw <- function(xyt,t.slice,vmax=NA){
  if (dim(xyt)[1] != 2){stop('The xyt dataframe does not contain 2 rows and 3 columns.')}
  if (dim(xyt)[2] != 3){stop('The xyt dataframe does not contain 2 rows and 3 columns.')}
  if (min(t.slice) < xyt[1,3]){stop('Invalid interpolation times in t.slice.')}
  if (max(t.slice) > xyt[2,3]){stop('Invalid interpolation times in t.slice.')}
  nameo <- names(xyt)
  xyt <- as.matrix(xyt)
  n <- length(t.slice)
  x1 <- xyt[1,1:2]
  x2 <- xyt[2,1:2]
  t1 <- xyt[1,3]
  t2 <- xyt[2,3]
  
  #Default value for vmax
  if(is.na(vmax)){
    vmax <- 1.5*( sqrt(sum((x2-x1)^2))/(t2-t1) )
  }

  prism.slice <- function(x1,x2,t1,t2,t.sli,vmax){
      #the number of points around the circle
      theta <- seq(0,2*pi,length.out=360)
      #calculate the radius of the Future Cone
      tF <- t.sli - t1
      rF <- vmax*tF 
      #calculate the radius of the Past Cone
      tP <- t2 - t.sli
      rP <- vmax*tP
      #get x,y coords of Future Cone circle
      xf <- x1[1] + rF*cos(theta)
      yf <- x1[2] + rF*sin(theta)
      #get x,y coords of Past Cone circle
      xp <- x2[1] + rP*cos(theta)
      yp <- x2[2] + rP*sin(theta)
      #create spatial polygons and intersect them
      c1 <- Polygon(rbind(cbind(xp,yp),c(xp[1],yp[1])))
      c2 <- Polygon(rbind(cbind(xf,yf),c(xf[1],yf[1])))
      spc1 <- SpatialPolygons(list(Polygons(list(c1),ID="1")))
      spc2 <- SpatialPolygons(list(Polygons(list(c2),ID="2")))
      slicePoly <- gIntersection(spc1,spc2) 
      return(slicePoly)
    }
  outdf <- data.frame(x=rep(NA,n), y=rep(NA,n), t=t.slice)
  outdf <- rbind(c(x1,t1),outdf,c(x2,t2))
  
  ti <- sample(1:n,n)
  for (i in 1:n){
    j <- ti[i]
    t. <- t.slice[j]
    
    ia <- max(which(!is.na(outdf$x) & outdf$t < t.))
    ib <- min(which(!is.na(outdf$x) & outdf$t > t.))
    xa <- c(outdf[ia,1],outdf[ia,2])
    xb <- c(outdf[ib,1],outdf[ib,2])
    
    reg <- prism.slice(xa,xb,outdf[ia,3],outdf[ib,3],t.,vmax)
    if (!is.null(reg)){
      sp. <- try(coordinates(spsample(reg,n=1,type='random'))[1,],TRUE)
      if (class(sp.) =='try-error'){sp. <- coordinates(gPointOnSurface(reg))}
    } else {
      sp. <- 0.5*(xa+xb)
    }
    
    #Save the coordinates
    jj <- which(outdf$t == t.)
    outdf[jj,1:2] <- sp.
  }
  names(outdf) <- nameo
  #don't return the anchor coordinates.
  return(outdf[2:(n+1),]) 
}


# x1 <- c(0,0)
# x2 <- c(100,100)
# t1 <- 0
# t2 <- 11
# t.slice <- 1:10
# vmax <- 25
# 
# tr <- tgrw(x1,x2,t1,t2,t.slice,vmax)
# 
# plot(tr[,1:2],xlim=c(-10,110),ylim=c(-10,110))
# points(x1[1],x1[2],pch=20)
# points(x2[1],x2[2],pch=20)


